src/OpenDB/src/lefout/lefout.cpp

```diff
@@ -32,80 +32,179 @@
+        via_name = box->getTechVia()->getName();
+      }
+      if (box->getBlockVia()) {
+        via_name = box->getBlockVia()->getName();
+      }
+
       int x, y;
       box->getViaXY(x, y);
-      std::string n = via->getName();
       fprintf(_out,
               "%sVIA %g %g %s ;\n",
               indent,
               lefdist(x),
               lefdist(y),
-              n.c_str());
+              via_name.c_str());
       cur_layer = NULL;
     } else {
-      int x1 = box->xMin();
-      int y1 = box->yMin();
-      int x2 = box->xMax();
-      int y2 = box->yMax();
-
-      std::string n;
+      std::string layer_name;
       if (_use_alias && layer->hasAlias())
-        n = layer->getAlias();
+        layer_name = layer->getAlias();
       else
-        n = layer->getName();
+        layer_name = layer->getName();
 
       if (cur_layer != layer) {
-        std::string n;
+        fprintf(_out, "%sLAYER %s ;\n", indent, layer_name.c_str());
+        cur_layer = layer;
+      }
+
+      writeBox(indent, box);
+    }
+  }
+}
 
-        if (_use_alias && layer->hasAlias())
-          n = layer->getAlias();
-        else
-          n = layer->getName();
+void lefout::writeBox(std::string indent, dbBox* box)
+{
+  int x1 = box->xMin();
+  int y1 = box->yMin();
+  int x2 = box->xMax();
+  int y2 = box->yMax();
 
-        fprintf(_out, "%sLAYER %s ;\n", indent, n.c_str());
-        cur_layer = layer;
+  fprintf(_out,
+          "%s  RECT  %g %g %g %g ;\n",
+          indent.c_str(),
+          lefdist(x1),
+          lefdist(y1),
+          lefdist(x2),
+          lefdist(y2));
+}
+
+void lefout::writeHeader(dbBlock* db_block)
+{
+  char left_bus_delimeter = 0;
+  char right_bus_delimeter = 0;
+  char hier_delimeter = db_block->getHierarchyDelimeter();
+
+  db_block->getBusDelimeters(left_bus_delimeter, right_bus_delimeter);
+
+  if (left_bus_delimeter == 0)
+    left_bus_delimeter = '[';
+
+  if (right_bus_delimeter == 0)
+    right_bus_delimeter = ']';
+
+  if (hier_delimeter == 0)
+    hier_delimeter = '|';
+
+  writeVersion("5.8");
+  writeBusBitChars(left_bus_delimeter, right_bus_delimeter);
+  writeDividerChar(hier_delimeter);
+}
+
+void lefout::writeObstructions(dbBlock* db_block)
+{
+  std::set<dbTechLayer*> obstruction_layers;
+  getTechLayerObstructions(db_block, obstruction_layers);
+
+  fprintf(_out, "  OBS\n");
+  dbBox* block_bounding_box = db_block->getBBox();
+  for (dbTechLayer* tech_layer : obstruction_layers) {
+    fprintf(_out, "    LAYER %s ;\n", tech_layer->getName().c_str());
+    writeBox("  ", block_bounding_box);
+  }
+  fprintf(_out, "  END\n");
+}
+
+void lefout::getTechLayerObstructions(
+    dbBlock* db_block,
+    std::set<dbTechLayer*>& obstruction_layers) const
+{
+  for (dbNet* net : db_block->getNets()) {
+    if (isPowerNet(net->getSigType())) {
+      continue;
+    }
+
+    findSWireLayerObstructions(obstruction_layers, net);
+    findWireLayerObstructions(obstruction_layers, net);
+  }
+}
```

Comment:
@arlpetergadfort @rovinski This is the code in question. Does this seem incorrect to you for any reason?

Reply:
yes, `isPowerNet` is checking that the signal is **not** a power net. therefore you are only getting OBS on layers with power nets on it.
The issue is also present in `writePowerPins` where you check if it's a power net.

Reply:
Just use

if (net->getSigType()->isSupply()) {
  continue;
}

Reply:
btw, why exclude power nets?

Reply:
My concern was that if the power nets don't extend to the end of the block and they reside on the top most layer they wouldn't be accessible.

I figured as long as long as there's only power nets on a layer it should be fine to allow pin access to them.

Reply:
See my other comment; it is okay to assume this will never happen. If you want to be thorough, you could issue a warning if a pin has no exterior connections possible (not drawn to edge and not on top layer).